{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0ee0179",
   "metadata": {},
   "outputs": [],
   "source": [
    "import anoexpress as xpress\n",
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "1886b971",
   "metadata": {},
   "source": [
    "### Genome-wide expression scans (GWES)\n",
    "\n",
    "In this notebook, we run genome-wide expression scans, plotting fold-changes for each gene on a contig with a moving average."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d91b57c2",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "def plot_contig_expression_track(\n",
    "    fold_change_df,\n",
    "    windowed_fold_change_df,\n",
    "    title=None, \n",
    "    width=800, \n",
    "    height=600, \n",
    "    sizing_mode='stretch_width',\n",
    "    x_range=None,\n",
    "    y_range=None,\n",
    "    show=False\n",
    "):\n",
    "    import bokeh\n",
    "    \n",
    "    # determine X axis range\n",
    "    x_min = fold_change_df.midpoint.to_numpy()[0]\n",
    "    x_max = fold_change_df.midpoint.to_numpy()[-1]\n",
    "    if x_range is None:\n",
    "        x_range = bokeh.models.Range1d(x_min, x_max, bounds=\"auto\")\n",
    "\n",
    "    # create a figure\n",
    "    xwheel_zoom = bokeh.models.WheelZoomTool(\n",
    "        dimensions=\"width\", maintain_focus=False\n",
    "    )\n",
    "    \n",
    "    tooltips = [\n",
    "            (\"Gene ID\", \"@GeneID\"),\n",
    "            (\"Gene Name\",\"@GeneName\"),\n",
    "            (\"Gene Description\", \"@GeneDescription\"),\n",
    "            (\"Experiment\", \"@comparison\"),\n",
    "            (\"Fold Change\", \"@fold_change\"),\n",
    "        ]\n",
    "\n",
    "    fig = bokeh.plotting.figure(\n",
    "        title=title,\n",
    "        tools=[\"xpan\", \"xzoom_in\", \"xzoom_out\", xwheel_zoom, \"reset\", \"hover\"],\n",
    "        active_scroll=xwheel_zoom,\n",
    "        active_drag=\"xpan\",\n",
    "        sizing_mode=sizing_mode,\n",
    "        width=width,\n",
    "        height=height,\n",
    "        toolbar_location=\"above\",\n",
    "        tooltips=tooltips,\n",
    "        x_range=x_range,\n",
    "        y_range=y_range\n",
    "    )\n",
    "\n",
    "    # plot \n",
    "    fig.circle(\n",
    "        x=\"midpoint\",\n",
    "        y=\"fold_change\",\n",
    "        size=4,\n",
    "        line_width=1,\n",
    "        line_color=\"grey\",\n",
    "        fill_color=None,\n",
    "        source=fold_change_df,\n",
    "    )\n",
    "    \n",
    "    fig.line(\n",
    "        x=\"midpoint\",\n",
    "        y=\"median_fc\",\n",
    "        line_width=2,\n",
    "        line_color=\"black\",\n",
    "        source=windowed_fold_change_df,\n",
    "    )\n",
    "\n",
    "    # tidy up the plot\n",
    "    fig.yaxis.axis_label = \"Log2 Fold Change\"\n",
    "\n",
    "    if show:\n",
    "        bokeh.plotting.show(fig)\n",
    "    return fig \n",
    "\n",
    "def plot_contig_expression(contig, analysis, data_type='fcs', microarray=False, pvalue_filter=None, y_range=(-10,15), height=400, width=600, title=None, size=10, step=None):\n",
    "    import bokeh\n",
    "    import malariagen_data\n",
    "    import allel\n",
    "    \n",
    "    fc_data = xpress.data(\n",
    "                        data_type=data_type, \n",
    "                        analysis=analysis, \n",
    "                        microarray=microarray, \n",
    "                        annotations=True, \n",
    "                        pvalue_filter=pvalue_filter\n",
    "                        ).reset_index()\n",
    "    \n",
    "    ag3 = malariagen_data.Ag3()\n",
    "    # calculate moving average\n",
    "    gff = ag3.genome_features(contig).query(\"type == 'gene'\").assign(midpoint = lambda x: (x.start + x.end)/2)\n",
    "    genes_df = gff[['ID', 'midpoint']].rename(columns={'ID':'GeneID'})\n",
    "    genes_df = genes_df.merge(fc_data, how='left').set_index(['GeneID', 'GeneName', 'GeneDescription', 'midpoint'])\n",
    "    median_fc = genes_df.mean(axis=1).to_frame().rename(columns={0:'median_fc'}).reset_index()\n",
    "    moving_mid = allel.moving_statistic(median_fc.midpoint, np.nanmedian, size=size, step=step)\n",
    "    moving_med = allel.moving_statistic(median_fc['median_fc'], np.nanmedian, size=size, step=step)\n",
    "    windowed_fold_change_df = pd.DataFrame({'midpoint':moving_mid, 'median_fc':moving_med})\n",
    "\n",
    "    fold_change_df = genes_df.reset_index().melt(id_vars=['midpoint', 'GeneID', 'GeneName', 'GeneDescription'], var_name='comparison', value_name='fold_change')  \n",
    "    \n",
    "    fig1 = plot_contig_expression_track(\n",
    "                                    df=fold_change_df,\n",
    "                                    moving_fc=windowed_fold_change_df,\n",
    "                                    y_range=y_range,\n",
    "                                    height=height,\n",
    "                                    width=width,\n",
    "                                    title=title,\n",
    "                                    show=False\n",
    "                                    )\n",
    "    fig1.xaxis.visible = False\n",
    "    # plot genes\n",
    "    fig2 = ag3.plot_genes(\n",
    "        region=contig,\n",
    "        sizing_mode=\"stretch_width\",\n",
    "        width=width,\n",
    "        height=100,\n",
    "        x_range=fig1.x_range,\n",
    "        show=False,\n",
    "    )\n",
    "    # combine plots into a single figure\n",
    "    fig = bokeh.layouts.gridplot(\n",
    "        [fig1, fig2],\n",
    "        ncols=1,\n",
    "        toolbar_location=\"above\",\n",
    "        merge_tools=True,\n",
    "        sizing_mode=\"stretch_width\",\n",
    "    )\n",
    "    bokeh.plotting.show(fig)\n",
    "    return fig\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "448d2ff7",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "plot_contig_expression(\n",
    "    contig='3RL',\n",
    "    analysis='gamb_colu', \n",
    "    microarray=False,\n",
    "    y_range=(-8,10), \n",
    "    size=10,\n",
    "    step=5,\n",
    "    height=400, \n",
    "    width=600)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
